{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ground Truth Data Extraction \n",
    "\n",
    "### Overview\n",
    "This notebook performs ground truth extraction from copy number variation (CNV) data generated from single-cell RNA sequencing (scRNA-seq).\n",
    "\n",
    "An overall collection of 30 patient samples are extracted using this notebook. \n",
    "\n",
    "The ground truth consists of copy number variations (CNVs) detected in tumour samples. Each row in this table represents a specific genomic segment with an altered copy number. It contains the following information:\n",
    "\n",
    "- GDC_Aliquot: This is a unique identifier for each tumour sample.\n",
    "- Chromosome: Indicates which chromosome the CNV is located on.\n",
    "- Start: The starting genomic position of the CNV segment.\n",
    "- End: The ending genomic position of the CNV segment.\n",
    "- Copy_Number: The total number of copies of this genomic segment in the tumour cells. In a normal diploid genome, this would be 2.\n",
    "- Major_Copy_Number: The number of copies of the more abundant allele.\n",
    "- Minor_Copy_Number: The number of copies of the less abundant allele."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_Aliquot</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Copy_Number</th>\n",
       "      <th>Major_Copy_Number</th>\n",
       "      <th>Minor_Copy_Number</th>\n",
       "      <th>Case_ID</th>\n",
       "      <th>Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2eb1196d-5e6d-4223-953c-d558c75a8c2e</td>\n",
       "      <td>chr1</td>\n",
       "      <td>13116</td>\n",
       "      <td>248945703</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C3L-00359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2eb1196d-5e6d-4223-953c-d558c75a8c2e</td>\n",
       "      <td>chr2</td>\n",
       "      <td>10587</td>\n",
       "      <td>242183243</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C3L-00359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2eb1196d-5e6d-4223-953c-d558c75a8c2e</td>\n",
       "      <td>chr3</td>\n",
       "      <td>18519</td>\n",
       "      <td>198181744</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C3L-00359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2eb1196d-5e6d-4223-953c-d558c75a8c2e</td>\n",
       "      <td>chr4</td>\n",
       "      <td>11961</td>\n",
       "      <td>190122722</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C3L-00359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2eb1196d-5e6d-4223-953c-d558c75a8c2e</td>\n",
       "      <td>chr5</td>\n",
       "      <td>11882</td>\n",
       "      <td>181363900</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>C3L-00359</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            GDC_Aliquot Chromosome  Start        End  \\\n",
       "0  2eb1196d-5e6d-4223-953c-d558c75a8c2e       chr1  13116  248945703   \n",
       "1  2eb1196d-5e6d-4223-953c-d558c75a8c2e       chr2  10587  242183243   \n",
       "2  2eb1196d-5e6d-4223-953c-d558c75a8c2e       chr3  18519  198181744   \n",
       "3  2eb1196d-5e6d-4223-953c-d558c75a8c2e       chr4  11961  190122722   \n",
       "4  2eb1196d-5e6d-4223-953c-d558c75a8c2e       chr5  11882  181363900   \n",
       "\n",
       "   Copy_Number  Major_Copy_Number  Minor_Copy_Number    Case_ID Sample  \n",
       "0            4                  2                  2  C3L-00359      1  \n",
       "1            4                  2                  2  C3L-00359      1  \n",
       "2            4                  2                  2  C3L-00359      1  \n",
       "3            4                  2                  2  C3L-00359      1  \n",
       "4            4                  2                  2  C3L-00359      1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Helper function to extract the ground truth files from the cluster directories\n",
    "def find_files_with_ending(root_dir, file_ending):\n",
    "    # List to store file details\n",
    "    file_data = []\n",
    "    \n",
    "    # Recursively run through each of the subdirectories\n",
    "    for dirpath, dirnames, filenames in os.walk(root_dir):\n",
    "        for file in filenames:\n",
    "            if file.endswith(file_ending):  # Match files with the specific ending\n",
    "                # Split the directory path into individual subdirectories\n",
    "                path_parts = dirpath.split(os.sep)\n",
    "                \n",
    "                file_info = {}\n",
    "                \n",
    "                # Excluding the root, by starting at 1, add subdirectory levels to the dictionary\n",
    "                for i, part in enumerate(path_parts[1:], start=1):\n",
    "                    file_info[f'subdir_{i}'] = part\n",
    "                \n",
    "                # Add the file path to the dictionary\n",
    "                file_info['file_path'] = os.path.join(dirpath, file)\n",
    "\n",
    "                file_data.append(file_info)\n",
    "    \n",
    "    # Convert the list of dictionaries to a DataFrame\n",
    "    df = pd.DataFrame(file_data)\n",
    "    return df\n",
    "\n",
    "# Function to read tab-delimited files from a DataFrame and combine them\n",
    "def read_tab_delimited_files_from_dataframe(df):\n",
    "    all_dataframes = []\n",
    "    for index, row in df.iterrows():\n",
    "        file_path = row['file_path']  # Get the full file path\n",
    "        try:\n",
    "            # Read the tab-delimited file\n",
    "            file_df = pd.read_csv(file_path, delimiter='\\t')\n",
    "            \n",
    "            # Include original metadata in the new DataFrame\n",
    "            for key in row.index:\n",
    "                if key not in file_df.columns and key != 'file_path':  # Avoid overwriting columns from the CSV and exclude 'file_path'\n",
    "                    file_df[key] = row[key]\n",
    "            \n",
    "            # Append the DataFrame to the list\n",
    "            all_dataframes.append(file_df)\n",
    "        except Exception as e:\n",
    "            print(f\"Error reading {file_path}: {e}\")\n",
    "    \n",
    "    # Combine all DataFrames into one\n",
    "    combined_df = pd.concat(all_dataframes, ignore_index=True)\n",
    "    \n",
    "    return combined_df\n",
    "\n",
    "root_directory = \"D:/GDC-data\"\n",
    "file_ending = \"copy_number_variation.seg.txt\"\n",
    "\n",
    "# Find files with the specified ending\n",
    "df_files = find_files_with_ending(root_directory, file_ending)\n",
    "\n",
    "# Read tab-delimited files from the DataFrame and combine them\n",
    "combined_dataframe = read_tab_delimited_files_from_dataframe(df_files)\n",
    "combined_dataframe.rename(columns={'subdir_1': 'Case_ID', 'subdir_2': 'Sample'}, inplace=True)\n",
    "combined_dataframe.drop(columns=['subdir_3', 'subdir_4'], inplace=True)\n",
    "display(combined_dataframe.head())\n",
    "\n",
    "# Export the DataFrame as a CSV file\n",
    "output_csv_path = \"ground_truth_combined.csv\"\n",
    "combined_dataframe.to_csv(output_csv_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
