{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Framework for Supervised Learning on CNV Data and Gene Expressions\n",
    "\n",
    "We generate a dataframe for one sample, seen below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ee5e869c-e15f-4899-9e12-377920609b42.wgs.ASCAT.gene_level.copy_number_variation.tsv has 60623 rows and 8 columns\n",
      "The number of unique gene_ids in the dataframe is 60623\n",
      "The number of unique gene_ids is equal to the number of rows for a sample\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id    gene_name chromosome  start    end  copy_number  \\\n",
       "0  ENSG00000223972.5      DDX11L1       chr1  11869  14409          4.0   \n",
       "1  ENSG00000227232.5       WASH7P       chr1  14404  29570          4.0   \n",
       "2  ENSG00000278267.1    MIR6859-1       chr1  17369  17436          4.0   \n",
       "3  ENSG00000243485.5  MIR1302-2HG       chr1  29554  31109          4.0   \n",
       "4  ENSG00000284332.1    MIR1302-2       chr1  30366  30503          4.0   \n",
       "\n",
       "   min_copy_number  max_copy_number  \n",
       "0              4.0              4.0  \n",
       "1              4.0              4.0  \n",
       "2              4.0              4.0  \n",
       "3              4.0              4.0  \n",
       "4              4.0              4.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_map = pd.read_csv('https://raw.githubusercontent.com/LaraLim/cnv-supervised-learning/refs/heads/main/sample_mapping.csv')\n",
    "\n",
    "\n",
    "#first sample \n",
    "filename = df_map['Filename'][0]\n",
    "\n",
    "#get the shape of the dataframe\n",
    "url = f'https://raw.githubusercontent.com/LaraLim/cnv-supervised-learning/refs/heads/main/gene_level_cnv/{filename}'\n",
    "df = pd.read_csv(url, sep='\\t')\n",
    "\n",
    "\n",
    "#find how many unique gene_ids are in the dataframe\n",
    "unique_genes = df['gene_id'].unique()\n",
    "num_genes = len(unique_genes)\n",
    "num_rows = df.shape[0]\n",
    "\n",
    "#print the number of rows and columns for a sample\n",
    "print(f'{filename} has {df.shape[0]} rows and {df.shape[1]} columns')\n",
    "\n",
    "print(f'The number of unique gene_ids in the dataframe is {num_genes}')\n",
    "\n",
    "# validate whether the number of unique gene_ids is equal to the number of rows in the dataframe\n",
    "if (num_genes == num_rows):\n",
    "    print(\"The number of unique gene_ids is equal to the number of rows for a sample\")\n",
    "\n",
    "display(df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we iterate over all of the samples, creating a dataframe for each and storing it into a dictionary identified by their sample identification.\n",
    "We want to ensure that each sample contains the same list of unique gene identifier and has the sample number of rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#dictionary of dataframes for each sample\n",
    "dfs = {}\n",
    "\n",
    "# loop through samples, creating a data frame for each sample\n",
    "for filename in df_map['Filename']:\n",
    "\n",
    "    #append new column for the case id\n",
    "    case_id = df_map[df_map['Filename'] == filename]['CaseID'].values[0]\n",
    "\n",
    "\n",
    "    url = f'https://raw.githubusercontent.com/LaraLim/cnv-supervised-learning/refs/heads/main/gene_level_cnv/{filename}'\n",
    "    df = pd.read_csv(url, sep='\\t')\n",
    "\n",
    "    #check if the number of rows is the same\n",
    "    if df.shape[0] != num_rows: \n",
    "        print(\"The number of rows is not the same between all files\")\n",
    "    \n",
    "    #check if the number of genes is the same\n",
    "    if len(df['gene_id'].unique()) != num_genes:\n",
    "        print(\"The number of genes is not the same between all files\")\n",
    "\n",
    "    #check whether the unique genes are the same\n",
    "    if not df['gene_id'].unique().all() == unique_genes.all():\n",
    "        print(\"The unique genes are not the same between all files\")\n",
    "    \n",
    "    df['CaseID'] = case_id\n",
    "    df['SumGeneExpression'] = 0\n",
    "    df['MeanGeneExpression'] = 0\n",
    "    df['VarianceGeneExpression'] = 0\n",
    "\n",
    "    df['status'] = 'normal'\n",
    "    df.loc[df['copy_number'] > 2, 'status'] = 'amplified'\n",
    "    df.loc[df['copy_number'] < 2, 'status'] = 'deleted'\n",
    "\n",
    "    #move the position of the copy number column to the last column\n",
    "    copy_number = df.pop('copy_number')\n",
    "    df['copy_number_target'] = copy_number\n",
    "\n",
    "    \n",
    "    #add the dataframe to the dictionary\n",
    "    dfs[case_id] = df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now know that the list of unique gene identifiers are consistent throughout the samples. We also know confirmed that the number unique gene identifiers for each sample is the same as the number of rows for each sample. This will allow us to perform transformations over the all the genes within each sample.\n",
    "\n",
    "\n",
    "### Data Imputations\n",
    "Now to check for NULL values for copy number variations, we check how many null values there are and if they are for the same genes throughout the samples. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>C3L-00359-01</th>\n",
       "      <th>C3L-00606-01</th>\n",
       "      <th>C3L-00606-02</th>\n",
       "      <th>C3L-00606-03</th>\n",
       "      <th>C3L-01287-01</th>\n",
       "      <th>C3L-01287-03</th>\n",
       "      <th>C3L-01953-01</th>\n",
       "      <th>C3L-02705-71</th>\n",
       "      <th>C3L-02858-01</th>\n",
       "      <th>C3L-03968-01</th>\n",
       "      <th>...</th>\n",
       "      <th>C3N-01904-02</th>\n",
       "      <th>C3N-02181-02</th>\n",
       "      <th>C3N-02188-03</th>\n",
       "      <th>C3N-02190-01-02</th>\n",
       "      <th>C3N-02769-02</th>\n",
       "      <th>C3N-02783-05</th>\n",
       "      <th>C3N-02784-01-03</th>\n",
       "      <th>C3N-03184-02</th>\n",
       "      <th>C3N-03186-01</th>\n",
       "      <th>C3N-03188-02</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chromosome</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min_copy_number</th>\n",
       "      <td>2225</td>\n",
       "      <td>2220</td>\n",
       "      <td>2217</td>\n",
       "      <td>2217</td>\n",
       "      <td>1709</td>\n",
       "      <td>1712</td>\n",
       "      <td>1709</td>\n",
       "      <td>1707</td>\n",
       "      <td>1716</td>\n",
       "      <td>2243</td>\n",
       "      <td>...</td>\n",
       "      <td>2220</td>\n",
       "      <td>2222</td>\n",
       "      <td>1709</td>\n",
       "      <td>1706</td>\n",
       "      <td>1714</td>\n",
       "      <td>2221</td>\n",
       "      <td>1721</td>\n",
       "      <td>1721</td>\n",
       "      <td>2226</td>\n",
       "      <td>1724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max_copy_number</th>\n",
       "      <td>2225</td>\n",
       "      <td>2220</td>\n",
       "      <td>2217</td>\n",
       "      <td>2217</td>\n",
       "      <td>1709</td>\n",
       "      <td>1712</td>\n",
       "      <td>1709</td>\n",
       "      <td>1707</td>\n",
       "      <td>1716</td>\n",
       "      <td>2243</td>\n",
       "      <td>...</td>\n",
       "      <td>2220</td>\n",
       "      <td>2222</td>\n",
       "      <td>1709</td>\n",
       "      <td>1706</td>\n",
       "      <td>1714</td>\n",
       "      <td>2221</td>\n",
       "      <td>1721</td>\n",
       "      <td>1721</td>\n",
       "      <td>2226</td>\n",
       "      <td>1724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CaseID</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SumGeneExpression</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MeanGeneExpression</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VarianceGeneExpression</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>status</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>copy_number_target</th>\n",
       "      <td>2225</td>\n",
       "      <td>2220</td>\n",
       "      <td>2217</td>\n",
       "      <td>2217</td>\n",
       "      <td>1709</td>\n",
       "      <td>1712</td>\n",
       "      <td>1709</td>\n",
       "      <td>1707</td>\n",
       "      <td>1716</td>\n",
       "      <td>2243</td>\n",
       "      <td>...</td>\n",
       "      <td>2220</td>\n",
       "      <td>2222</td>\n",
       "      <td>1709</td>\n",
       "      <td>1706</td>\n",
       "      <td>1714</td>\n",
       "      <td>2221</td>\n",
       "      <td>1721</td>\n",
       "      <td>1721</td>\n",
       "      <td>2226</td>\n",
       "      <td>1724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        C3L-00359-01  C3L-00606-01  C3L-00606-02  \\\n",
       "gene_id                            0             0             0   \n",
       "gene_name                          0             0             0   \n",
       "chromosome                         0             0             0   \n",
       "start                              0             0             0   \n",
       "end                                0             0             0   \n",
       "min_copy_number                 2225          2220          2217   \n",
       "max_copy_number                 2225          2220          2217   \n",
       "CaseID                             0             0             0   \n",
       "SumGeneExpression                  0             0             0   \n",
       "MeanGeneExpression                 0             0             0   \n",
       "VarianceGeneExpression             0             0             0   \n",
       "status                             0             0             0   \n",
       "copy_number_target              2225          2220          2217   \n",
       "\n",
       "                        C3L-00606-03  C3L-01287-01  C3L-01287-03  \\\n",
       "gene_id                            0             0             0   \n",
       "gene_name                          0             0             0   \n",
       "chromosome                         0             0             0   \n",
       "start                              0             0             0   \n",
       "end                                0             0             0   \n",
       "min_copy_number                 2217          1709          1712   \n",
       "max_copy_number                 2217          1709          1712   \n",
       "CaseID                             0             0             0   \n",
       "SumGeneExpression                  0             0             0   \n",
       "MeanGeneExpression                 0             0             0   \n",
       "VarianceGeneExpression             0             0             0   \n",
       "status                             0             0             0   \n",
       "copy_number_target              2217          1709          1712   \n",
       "\n",
       "                        C3L-01953-01  C3L-02705-71  C3L-02858-01  \\\n",
       "gene_id                            0             0             0   \n",
       "gene_name                          0             0             0   \n",
       "chromosome                         0             0             0   \n",
       "start                              0             0             0   \n",
       "end                                0             0             0   \n",
       "min_copy_number                 1709          1707          1716   \n",
       "max_copy_number                 1709          1707          1716   \n",
       "CaseID                             0             0             0   \n",
       "SumGeneExpression                  0             0             0   \n",
       "MeanGeneExpression                 0             0             0   \n",
       "VarianceGeneExpression             0             0             0   \n",
       "status                             0             0             0   \n",
       "copy_number_target              1709          1707          1716   \n",
       "\n",
       "                        C3L-03968-01  ...  C3N-01904-02  C3N-02181-02  \\\n",
       "gene_id                            0  ...             0             0   \n",
       "gene_name                          0  ...             0             0   \n",
       "chromosome                         0  ...             0             0   \n",
       "start                              0  ...             0             0   \n",
       "end                                0  ...             0             0   \n",
       "min_copy_number                 2243  ...          2220          2222   \n",
       "max_copy_number                 2243  ...          2220          2222   \n",
       "CaseID                             0  ...             0             0   \n",
       "SumGeneExpression                  0  ...             0             0   \n",
       "MeanGeneExpression                 0  ...             0             0   \n",
       "VarianceGeneExpression             0  ...             0             0   \n",
       "status                             0  ...             0             0   \n",
       "copy_number_target              2243  ...          2220          2222   \n",
       "\n",
       "                        C3N-02188-03  C3N-02190-01-02  C3N-02769-02  \\\n",
       "gene_id                            0                0             0   \n",
       "gene_name                          0                0             0   \n",
       "chromosome                         0                0             0   \n",
       "start                              0                0             0   \n",
       "end                                0                0             0   \n",
       "min_copy_number                 1709             1706          1714   \n",
       "max_copy_number                 1709             1706          1714   \n",
       "CaseID                             0                0             0   \n",
       "SumGeneExpression                  0                0             0   \n",
       "MeanGeneExpression                 0                0             0   \n",
       "VarianceGeneExpression             0                0             0   \n",
       "status                             0                0             0   \n",
       "copy_number_target              1709             1706          1714   \n",
       "\n",
       "                        C3N-02783-05  C3N-02784-01-03  C3N-03184-02  \\\n",
       "gene_id                            0                0             0   \n",
       "gene_name                          0                0             0   \n",
       "chromosome                         0                0             0   \n",
       "start                              0                0             0   \n",
       "end                                0                0             0   \n",
       "min_copy_number                 2221             1721          1721   \n",
       "max_copy_number                 2221             1721          1721   \n",
       "CaseID                             0                0             0   \n",
       "SumGeneExpression                  0                0             0   \n",
       "MeanGeneExpression                 0                0             0   \n",
       "VarianceGeneExpression             0                0             0   \n",
       "status                             0                0             0   \n",
       "copy_number_target              2221             1721          1721   \n",
       "\n",
       "                        C3N-03186-01  C3N-03188-02  \n",
       "gene_id                            0             0  \n",
       "gene_name                          0             0  \n",
       "chromosome                         0             0  \n",
       "start                              0             0  \n",
       "end                                0             0  \n",
       "min_copy_number                 2226          1724  \n",
       "max_copy_number                 2226          1724  \n",
       "CaseID                             0             0  \n",
       "SumGeneExpression                  0             0  \n",
       "MeanGeneExpression                 0             0  \n",
       "VarianceGeneExpression             0             0  \n",
       "status                             0             0  \n",
       "copy_number_target              2226          1724  \n",
       "\n",
       "[13 rows x 37 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>ENSG00000284740.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>ENSG00000286989.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>ENSG00000284745.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>ENSG00000284668.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>ENSG00000284703.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60618</th>\n",
       "      <td>ENSG00000124334.17_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60619</th>\n",
       "      <td>ENSG00000270726.6_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60620</th>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60621</th>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60622</th>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2225 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id\n",
       "119           ENSG00000284740.1\n",
       "126           ENSG00000286989.1\n",
       "180           ENSG00000284745.1\n",
       "213           ENSG00000284668.2\n",
       "214           ENSG00000284703.1\n",
       "...                         ...\n",
       "60618  ENSG00000124334.17_PAR_Y\n",
       "60619   ENSG00000270726.6_PAR_Y\n",
       "60620  ENSG00000185203.12_PAR_Y\n",
       "60621  ENSG00000182484.15_PAR_Y\n",
       "60622   ENSG00000227159.8_PAR_Y\n",
       "\n",
       "[2225 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Now to check for NULL values for copy number variations, we check how many null values there are and if they are for the same genes throughout the samples. \n",
    "null_values = pd.DataFrame()\n",
    "\n",
    "for case_id, df in dfs.items():\n",
    "    null_values[case_id] = df.isnull().sum()\n",
    "\n",
    "display(null_values)\n",
    "\n",
    "\n",
    "# for each sample in the dictionary, find the gene associated for a copy number that is null and check if they are the same genes for all samples\n",
    "null_genes = {}\n",
    "for case_id, df in dfs.items():\n",
    "    df  = pd.DataFrame()\n",
    "    df['gene_id'] = dfs[case_id][dfs[case_id]['copy_number_target'].isnull()]['gene_id']\n",
    "    null_genes[case_id] = df\n",
    "\n",
    "# for case_id, df in null_genes.items():\n",
    "#     print(f'Null genes for {case_id}')\n",
    "#     display(df)\n",
    "\n",
    "#display the first sample's null genes\n",
    "display(null_genes['C3L-00359-01'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the the copy_number_target is null for a varying number of occurrances. However it remains **mostly** consistent for the selection of genes throughout the samples. Therefore, to handle these missing values while remaining consistent with our datasets, we will impute the copy numbers with a negative one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "      <th>CaseID</th>\n",
       "      <th>SumGeneExpression</th>\n",
       "      <th>MeanGeneExpression</th>\n",
       "      <th>VarianceGeneExpression</th>\n",
       "      <th>status</th>\n",
       "      <th>copy_number_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60618</th>\n",
       "      <td>ENSG00000124334.17_PAR_Y</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57184101</td>\n",
       "      <td>57197337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60619</th>\n",
       "      <td>ENSG00000270726.6_PAR_Y</td>\n",
       "      <td>AJ271736.1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60620</th>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60621</th>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60622</th>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id   gene_name chromosome     start       end  \\\n",
       "60618  ENSG00000124334.17_PAR_Y        IL9R       chrY  57184101  57197337   \n",
       "60619   ENSG00000270726.6_PAR_Y  AJ271736.1       chrY  57190738  57208756   \n",
       "60620  ENSG00000185203.12_PAR_Y      WASIR1       chrY  57201143  57203357   \n",
       "60621  ENSG00000182484.15_PAR_Y      WASH6P       chrY  57207346  57212230   \n",
       "60622   ENSG00000227159.8_PAR_Y    DDX11L16       chrY  57212184  57214397   \n",
       "\n",
       "       min_copy_number  max_copy_number        CaseID  SumGeneExpression  \\\n",
       "60618              NaN              NaN  C3L-00359-01                  0   \n",
       "60619              NaN              NaN  C3L-00359-01                  0   \n",
       "60620              NaN              NaN  C3L-00359-01                  0   \n",
       "60621              NaN              NaN  C3L-00359-01                  0   \n",
       "60622              NaN              NaN  C3L-00359-01                  0   \n",
       "\n",
       "       MeanGeneExpression  VarianceGeneExpression  status  copy_number_target  \n",
       "60618                   0                       0  normal                -1.0  \n",
       "60619                   0                       0  normal                -1.0  \n",
       "60620                   0                       0  normal                -1.0  \n",
       "60621                   0                       0  normal                -1.0  \n",
       "60622                   0                       0  normal                -1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#impute the copy number values that are missing with -1 \n",
    "for case_id, df in dfs.items():\n",
    "    dfs[case_id]['copy_number_target'] = dfs[case_id]['copy_number_target'].fillna(-1)\n",
    "\n",
    "display(dfs['C3L-00359-01'].tail())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From processing the statistical data on Gene Expressions, we also know that the dense matrices do not exist on the cluster for the following cases and their samples:\n",
    "\n",
    "- C3N-01180\n",
    "- C3N-01334\n",
    "\n",
    "Therefore, we are removing these samples from our dataset.\n",
    "\n",
    "We are left with a total number of 36 samples. \n",
    "\n",
    "\n",
    "Next, we drop the CaseID, gene_id, gene_name, min_copy_number, max_copy_number, status (for the regression problem). We would keep the status for a classification problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "      <th>CaseID</th>\n",
       "      <th>SumGeneExpression</th>\n",
       "      <th>MeanGeneExpression</th>\n",
       "      <th>VarianceGeneExpression</th>\n",
       "      <th>status</th>\n",
       "      <th>copy_number_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60618</th>\n",
       "      <td>ENSG00000124334.17_PAR_Y</td>\n",
       "      <td>IL9R</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57184101</td>\n",
       "      <td>57197337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60619</th>\n",
       "      <td>ENSG00000270726.6_PAR_Y</td>\n",
       "      <td>AJ271736.1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57190738</td>\n",
       "      <td>57208756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60620</th>\n",
       "      <td>ENSG00000185203.12_PAR_Y</td>\n",
       "      <td>WASIR1</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57201143</td>\n",
       "      <td>57203357</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60621</th>\n",
       "      <td>ENSG00000182484.15_PAR_Y</td>\n",
       "      <td>WASH6P</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57207346</td>\n",
       "      <td>57212230</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60622</th>\n",
       "      <td>ENSG00000227159.8_PAR_Y</td>\n",
       "      <td>DDX11L16</td>\n",
       "      <td>chrY</td>\n",
       "      <td>57212184</td>\n",
       "      <td>57214397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C3N-01334-03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60623 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_id    gene_name chromosome     start       end  \\\n",
       "0             ENSG00000223972.5      DDX11L1       chr1     11869     14409   \n",
       "1             ENSG00000227232.5       WASH7P       chr1     14404     29570   \n",
       "2             ENSG00000278267.1    MIR6859-1       chr1     17369     17436   \n",
       "3             ENSG00000243485.5  MIR1302-2HG       chr1     29554     31109   \n",
       "4             ENSG00000284332.1    MIR1302-2       chr1     30366     30503   \n",
       "...                         ...          ...        ...       ...       ...   \n",
       "60618  ENSG00000124334.17_PAR_Y         IL9R       chrY  57184101  57197337   \n",
       "60619   ENSG00000270726.6_PAR_Y   AJ271736.1       chrY  57190738  57208756   \n",
       "60620  ENSG00000185203.12_PAR_Y       WASIR1       chrY  57201143  57203357   \n",
       "60621  ENSG00000182484.15_PAR_Y       WASH6P       chrY  57207346  57212230   \n",
       "60622   ENSG00000227159.8_PAR_Y     DDX11L16       chrY  57212184  57214397   \n",
       "\n",
       "       min_copy_number  max_copy_number        CaseID  SumGeneExpression  \\\n",
       "0                  4.0              4.0  C3N-01334-03                  0   \n",
       "1                  4.0              4.0  C3N-01334-03                  0   \n",
       "2                  4.0              4.0  C3N-01334-03                  0   \n",
       "3                  4.0              4.0  C3N-01334-03                  0   \n",
       "4                  4.0              4.0  C3N-01334-03                  0   \n",
       "...                ...              ...           ...                ...   \n",
       "60618              NaN              NaN  C3N-01334-03                  0   \n",
       "60619              NaN              NaN  C3N-01334-03                  0   \n",
       "60620              NaN              NaN  C3N-01334-03                  0   \n",
       "60621              NaN              NaN  C3N-01334-03                  0   \n",
       "60622              NaN              NaN  C3N-01334-03                  0   \n",
       "\n",
       "       MeanGeneExpression  VarianceGeneExpression     status  \\\n",
       "0                       0                       0  amplified   \n",
       "1                       0                       0  amplified   \n",
       "2                       0                       0  amplified   \n",
       "3                       0                       0  amplified   \n",
       "4                       0                       0  amplified   \n",
       "...                   ...                     ...        ...   \n",
       "60618                   0                       0     normal   \n",
       "60619                   0                       0     normal   \n",
       "60620                   0                       0     normal   \n",
       "60621                   0                       0     normal   \n",
       "60622                   0                       0     normal   \n",
       "\n",
       "       copy_number_target  \n",
       "0                     4.0  \n",
       "1                     4.0  \n",
       "2                     4.0  \n",
       "3                     4.0  \n",
       "4                     4.0  \n",
       "...                   ...  \n",
       "60618                -1.0  \n",
       "60619                -1.0  \n",
       "60620                -1.0  \n",
       "60621                -1.0  \n",
       "60622                -1.0  \n",
       "\n",
       "[60623 rows x 13 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove the C3N-01180 case ids from the dictionary\n",
    "dfs.pop('C3N-01180-01')\n",
    "dfs.pop('C3N-01334-03')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transform Numerical Features\n",
    "\n",
    "For each sample, we transform the numerical features: \n",
    "- start position (normalize)\n",
    "- end position (normalize)\n",
    "- Sum Gene Expression (lognormal)\n",
    "- Mean Gene Expression (lognormal)\n",
    "- Variance Gene Expression (lognormal)\n",
    "\n",
    "Example of normalizing the gene expressions for a few samples: https://github.com/LaraLim/cnv-supervised-learning/blob/main/visualize_genes_per_sample.ipynb\n",
    "\n",
    "\n",
    "**TODO:** ask Ted before we log normalize the all of the genes expressions per sample, whether we should ***also*** perform any transformations of the gene accross samples\n",
    "https://github.com/LaraLim/cnv-supervised-learning/blob/main/visualize_genes_across_samples.ipynb\n",
    "\n",
    "\n",
    "### Combining Samples For ML\n",
    "After we transform the numerical features for each sample, we combine the samples for ML. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "      <th>CaseID</th>\n",
       "      <th>SumGeneExpression</th>\n",
       "      <th>MeanGeneExpression</th>\n",
       "      <th>VarianceGeneExpression</th>\n",
       "      <th>status</th>\n",
       "      <th>copy_number_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>C3L-00359-01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id    gene_name chromosome  start    end  min_copy_number  \\\n",
       "0  ENSG00000223972.5      DDX11L1       chr1  11869  14409              4.0   \n",
       "1  ENSG00000227232.5       WASH7P       chr1  14404  29570              4.0   \n",
       "2  ENSG00000278267.1    MIR6859-1       chr1  17369  17436              4.0   \n",
       "3  ENSG00000243485.5  MIR1302-2HG       chr1  29554  31109              4.0   \n",
       "4  ENSG00000284332.1    MIR1302-2       chr1  30366  30503              4.0   \n",
       "\n",
       "   max_copy_number        CaseID  SumGeneExpression  MeanGeneExpression  \\\n",
       "0              4.0  C3L-00359-01                  0                   0   \n",
       "1              4.0  C3L-00359-01                  0                   0   \n",
       "2              4.0  C3L-00359-01                  0                   0   \n",
       "3              4.0  C3L-00359-01                  0                   0   \n",
       "4              4.0  C3L-00359-01                  0                   0   \n",
       "\n",
       "   VarianceGeneExpression     status  copy_number_target  \n",
       "0                       0  amplified                 4.0  \n",
       "1                       0  amplified                 4.0  \n",
       "2                       0  amplified                 4.0  \n",
       "3                       0  amplified                 4.0  \n",
       "4                       0  amplified                 4.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "combined_df = pd.concat(dfs.values(), ignore_index=True)\n",
    "display(combined_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The combined dataframe has 2303674 rows and 7 columns\n"
     ]
    }
   ],
   "source": [
    "#display dimensions of the combined dataframe\n",
    "print(f'The combined dataframe has {combined_df.shape[0]} rows and {combined_df.shape[1]} columns')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop the Columns for Gene Identifiers & CaseID\n",
    "\n",
    "We don't want the ML model to correlate cnv to a specific identifer, such as\n",
    "- gene_id\n",
    "- gene_name\n",
    "- CaseID\n",
    "\n",
    "\n",
    "For a regression problem we drop the status. \n",
    "**TODO** ask Ted if we should also consider comparing to a classification problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop cols from the combined dataframe\n",
    "combined_df.drop(['gene_id', 'gene_name', 'CaseID' 'status'], axis=1, inplace=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Split the data for machine learning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#split for training and testing\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "X = combined_df.drop('copy_number_target', axis=1)\n",
    "y = combined_df['copy_number_target']\n",
    "\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "#train SVM, Random Forest, etc, comparing performance\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***TODO***\n",
    "Ask Ted if we will need to perform any further transformations on the numerical data training and test data after the split given that we already have normalized values for each sample. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
