{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n",
      "Index([], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "# Create a dictionary to store the dataframes\n",
    "data = {}\n",
    "\n",
    "# Iterate through all the csv files in the dataset directory and store them in a dictionary with the key as the filename\n",
    "dataset_directory = '../dataset'\n",
    "for filename in os.listdir(dataset_directory):\n",
    "    if filename.endswith('.csv'):\n",
    "        file_path = os.path.join(dataset_directory, filename)\n",
    "        # Remove .csv from the filename\n",
    "        filename = filename[:-4]\n",
    "        df = pd.read_csv(file_path)\n",
    "      \n",
    "        df['sum_gene_expr_normalized'] = np.log1p(df['sum_gene_expr'])\n",
    "        df['mean_gene_expr_normalized'] = np.log1p(df['mean_gene_expr'])\n",
    "        df['variance_gene_expr_normalized'] = np.log1p(df['variance_gene_expr'])\n",
    "\n",
    "        #for the sum gene expression, create a window of 100 to store the avg stats\n",
    "        window_size = 100\n",
    "        df['window_sum_gene_expr_avg'] = df['sum_gene_expr'].rolling(window=window_size, min_periods=1).mean()\n",
    "        df['window_mean_gene_expr_avg'] = df['mean_gene_expr'].rolling(window=window_size, min_periods=1).mean()\n",
    "        df['window_variance_gene_expr_avg'] = df['variance_gene_expr'].rolling(window=window_size, min_periods=1).mean()\n",
    "\n",
    "        # Store the DataFrame in the dictionary\n",
    "        data[filename] = df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CaseID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "      <th>status</th>\n",
       "      <th>sum_gene_expr</th>\n",
       "      <th>mean_gene_expr</th>\n",
       "      <th>variance_gene_expr</th>\n",
       "      <th>copy_number_target</th>\n",
       "      <th>sum_gene_expr_normalized</th>\n",
       "      <th>mean_gene_expr_normalized</th>\n",
       "      <th>variance_gene_expr_normalized</th>\n",
       "      <th>window_sum_gene_expr_avg</th>\n",
       "      <th>window_mean_gene_expr_avg</th>\n",
       "      <th>window_variance_gene_expr_avg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.006084</td>\n",
       "      <td>4.167978e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>4.167977e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>8.853096e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>5.902064e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.002205</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>5.353520e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>4.282816e-11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         CaseID            gene_id    gene_name chromosome  start    end  \\\n",
       "0  C3L-00606-01  ENSG00000223972.5      DDX11L1       chr1  11869  14409   \n",
       "1  C3L-00606-01  ENSG00000227232.5       WASH7P       chr1  14404  29570   \n",
       "2  C3L-00606-01  ENSG00000278267.1    MIR6859-1       chr1  17369  17436   \n",
       "3  C3L-00606-01  ENSG00000243485.5  MIR1302-2HG       chr1  29554  31109   \n",
       "4  C3L-00606-01  ENSG00000284332.1    MIR1302-2       chr1  30366  30503   \n",
       "\n",
       "   min_copy_number  max_copy_number     status  sum_gene_expr  mean_gene_expr  \\\n",
       "0              4.0              4.0  amplified       0.000000    0.000000e+00   \n",
       "1              4.0              4.0  amplified       0.006084    4.167978e-07   \n",
       "2              4.0              4.0  amplified       0.000000    0.000000e+00   \n",
       "3              4.0              4.0  amplified       0.002205    1.510508e-07   \n",
       "4              4.0              4.0  amplified       0.000000    0.000000e+00   \n",
       "\n",
       "   variance_gene_expr  copy_number_target  sum_gene_expr_normalized  \\\n",
       "0        0.000000e+00                 4.0                  0.000000   \n",
       "1        1.770619e-10                 4.0                  0.006066   \n",
       "2        0.000000e+00                 4.0                  0.000000   \n",
       "3        3.707889e-11                 4.0                  0.002203   \n",
       "4        0.000000e+00                 4.0                  0.000000   \n",
       "\n",
       "   mean_gene_expr_normalized  variance_gene_expr_normalized  \\\n",
       "0               0.000000e+00                   0.000000e+00   \n",
       "1               4.167977e-07                   1.770619e-10   \n",
       "2               0.000000e+00                   0.000000e+00   \n",
       "3               1.510508e-07                   3.707889e-11   \n",
       "4               0.000000e+00                   0.000000e+00   \n",
       "\n",
       "   window_sum_gene_expr_avg  window_mean_gene_expr_avg  \\\n",
       "0                  0.000000               0.000000e+00   \n",
       "1                  0.003042               2.083989e-07   \n",
       "2                  0.002028               1.389326e-07   \n",
       "3                  0.002072               1.419621e-07   \n",
       "4                  0.001658               1.135697e-07   \n",
       "\n",
       "   window_variance_gene_expr_avg  \n",
       "0                   0.000000e+00  \n",
       "1                   8.853096e-11  \n",
       "2                   5.902064e-11  \n",
       "3                   5.353520e-11  \n",
       "4                   4.282816e-11  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# display a dataframe from the dictionary\n",
    "display(data['C3L-00359-01'].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CaseID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>min_copy_number</th>\n",
       "      <th>max_copy_number</th>\n",
       "      <th>status</th>\n",
       "      <th>sum_gene_expr</th>\n",
       "      <th>...</th>\n",
       "      <th>copy_number_target</th>\n",
       "      <th>sum_gene_expr_normalized</th>\n",
       "      <th>mean_gene_expr_normalized</th>\n",
       "      <th>variance_gene_expr_normalized</th>\n",
       "      <th>window_sum_gene_expr_avg</th>\n",
       "      <th>window_mean_gene_expr_avg</th>\n",
       "      <th>window_variance_gene_expr_avg</th>\n",
       "      <th>dev_gene_expr_normalized_mean</th>\n",
       "      <th>dev_mean_gene_expr_normalized_mean</th>\n",
       "      <th>dev_variance_gene_expr_normalized_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>11869.0</td>\n",
       "      <td>14409.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>chr1</td>\n",
       "      <td>14404.0</td>\n",
       "      <td>29570.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.006084</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>4.167977e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>8.853096e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>17369.0</td>\n",
       "      <td>17436.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>5.902064e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>chr1</td>\n",
       "      <td>29554.0</td>\n",
       "      <td>31109.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.002205</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>5.353520e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-00606-01</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>30366.0</td>\n",
       "      <td>30503.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>4.282816e-11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CaseID            gene_id    gene_name chromosome    start      end  \\\n",
       "0  C3L-00606-01  ENSG00000223972.5      DDX11L1       chr1  11869.0  14409.0   \n",
       "1  C3L-00606-01  ENSG00000227232.5       WASH7P       chr1  14404.0  29570.0   \n",
       "2  C3L-00606-01  ENSG00000278267.1    MIR6859-1       chr1  17369.0  17436.0   \n",
       "3  C3L-00606-01  ENSG00000243485.5  MIR1302-2HG       chr1  29554.0  31109.0   \n",
       "4  C3L-00606-01  ENSG00000284332.1    MIR1302-2       chr1  30366.0  30503.0   \n",
       "\n",
       "   min_copy_number  max_copy_number     status  sum_gene_expr  ...  \\\n",
       "0              4.0              4.0  amplified       0.000000  ...   \n",
       "1              4.0              4.0  amplified       0.006084  ...   \n",
       "2              4.0              4.0  amplified       0.000000  ...   \n",
       "3              4.0              4.0  amplified       0.002205  ...   \n",
       "4              4.0              4.0  amplified       0.000000  ...   \n",
       "\n",
       "   copy_number_target  sum_gene_expr_normalized  mean_gene_expr_normalized  \\\n",
       "0                 4.0                  0.000000               0.000000e+00   \n",
       "1                 4.0                  0.006066               4.167977e-07   \n",
       "2                 4.0                  0.000000               0.000000e+00   \n",
       "3                 4.0                  0.002203               1.510508e-07   \n",
       "4                 4.0                  0.000000               0.000000e+00   \n",
       "\n",
       "   variance_gene_expr_normalized  window_sum_gene_expr_avg  \\\n",
       "0                   0.000000e+00                  0.000000   \n",
       "1                   1.770619e-10                  0.003042   \n",
       "2                   0.000000e+00                  0.002028   \n",
       "3                   3.707889e-11                  0.002072   \n",
       "4                   0.000000e+00                  0.001658   \n",
       "\n",
       "   window_mean_gene_expr_avg  window_variance_gene_expr_avg  \\\n",
       "0               0.000000e+00                   0.000000e+00   \n",
       "1               2.083989e-07                   8.853096e-11   \n",
       "2               1.389326e-07                   5.902064e-11   \n",
       "3               1.419621e-07                   5.353520e-11   \n",
       "4               1.135697e-07                   4.282816e-11   \n",
       "\n",
       "   dev_gene_expr_normalized_mean  dev_mean_gene_expr_normalized_mean  \\\n",
       "0                            NaN                                 NaN   \n",
       "1                            NaN                                 NaN   \n",
       "2                            NaN                                 NaN   \n",
       "3                            NaN                                 NaN   \n",
       "4                            NaN                                 NaN   \n",
       "\n",
       "   dev_variance_gene_expr_normalized_mean  \n",
       "0                                     NaN  \n",
       "1                                     NaN  \n",
       "2                                     NaN  \n",
       "3                                     NaN  \n",
       "4                                     NaN  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2169985, 22)\n"
     ]
    }
   ],
   "source": [
    "combined_data = pd.concat(data.values(), axis=0)\n",
    "\n",
    "#display the combined data\n",
    "display(combined_data.head())\n",
    "\n",
    "print(combined_data.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58762\n",
      "58730\n",
      "24\n",
      "58918\n"
     ]
    }
   ],
   "source": [
    "#one hot encode the chromosomes\n",
    "combined_data = pd.get_dummies(combined_data, columns=['chromosome'])\n",
    "\n",
    "#drop the gene_id column\n",
    "combined_data = combined_data.drop('gene_id', axis=1)\n",
    "\n",
    "#drop the CaseID column\n",
    "combined_data = combined_data.drop('CaseID', axis=1)\n",
    "\n",
    "#drop the start and end columns\n",
    "combined_data = combined_data.drop('start', axis=1)\n",
    "combined_data = combined_data.drop('end', axis=1)\n",
    "\n",
    "#drop the min, max, and copy number targets\n",
    "combined_data = combined_data.drop('min_copy_number', axis=1)\n",
    "combined_data = combined_data.drop('max_copy_number', axis=1)\n",
    "combined_data = combined_data.drop('copy_number_target', axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New feature:\n",
    "- mean gene expressions - across cells but within one patient\n",
    "- take sume, mean, and variance gene expression after log transform then further subtract from it the overall means of the sum, mean, or variance of that gene across samples \n",
    "- This represents the mean gene expression across cells from this patient compared to other patients. We are getting each row's deviation from the gene's sum, mean, and variance\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>status</th>\n",
       "      <th>sum_gene_expr</th>\n",
       "      <th>mean_gene_expr</th>\n",
       "      <th>variance_gene_expr</th>\n",
       "      <th>sum_gene_expr_normalized</th>\n",
       "      <th>mean_gene_expr_normalized</th>\n",
       "      <th>variance_gene_expr_normalized</th>\n",
       "      <th>window_sum_gene_expr_avg</th>\n",
       "      <th>window_mean_gene_expr_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>chromosome_chr22</th>\n",
       "      <th>chromosome_chr3</th>\n",
       "      <th>chromosome_chr4</th>\n",
       "      <th>chromosome_chr5</th>\n",
       "      <th>chromosome_chr6</th>\n",
       "      <th>chromosome_chr7</th>\n",
       "      <th>chromosome_chr8</th>\n",
       "      <th>chromosome_chr9</th>\n",
       "      <th>chromosome_chrX</th>\n",
       "      <th>chromosome_chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.006084</td>\n",
       "      <td>4.167978e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>4.167977e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.002205</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 238 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name     status  sum_gene_expr  mean_gene_expr  variance_gene_expr  \\\n",
       "0      DDX11L1  amplified       0.000000    0.000000e+00        0.000000e+00   \n",
       "1       WASH7P  amplified       0.006084    4.167978e-07        1.770619e-10   \n",
       "2    MIR6859-1  amplified       0.000000    0.000000e+00        0.000000e+00   \n",
       "3  MIR1302-2HG  amplified       0.002205    1.510508e-07        3.707889e-11   \n",
       "4    MIR1302-2  amplified       0.000000    0.000000e+00        0.000000e+00   \n",
       "\n",
       "   sum_gene_expr_normalized  mean_gene_expr_normalized  \\\n",
       "0                  0.000000               0.000000e+00   \n",
       "1                  0.006066               4.167977e-07   \n",
       "2                  0.000000               0.000000e+00   \n",
       "3                  0.002203               1.510508e-07   \n",
       "4                  0.000000               0.000000e+00   \n",
       "\n",
       "   variance_gene_expr_normalized  window_sum_gene_expr_avg  \\\n",
       "0                   0.000000e+00                  0.000000   \n",
       "1                   1.770619e-10                  0.003042   \n",
       "2                   0.000000e+00                  0.002028   \n",
       "3                   3.707889e-11                  0.002072   \n",
       "4                   0.000000e+00                  0.001658   \n",
       "\n",
       "   window_mean_gene_expr_avg  ...  chromosome_chr22  chromosome_chr3  \\\n",
       "0               0.000000e+00  ...             False            False   \n",
       "1               2.083989e-07  ...             False            False   \n",
       "2               1.389326e-07  ...             False            False   \n",
       "3               1.419621e-07  ...             False            False   \n",
       "4               1.135697e-07  ...             False            False   \n",
       "\n",
       "   chromosome_chr4  chromosome_chr5  chromosome_chr6  chromosome_chr7  \\\n",
       "0            False            False            False            False   \n",
       "1            False            False            False            False   \n",
       "2            False            False            False            False   \n",
       "3            False            False            False            False   \n",
       "4            False            False            False            False   \n",
       "\n",
       "   chromosome_chr8  chromosome_chr9  chromosome_chrX  chromosome_chrY  \n",
       "0            False            False            False            False  \n",
       "1            False            False            False            False  \n",
       "2            False            False            False            False  \n",
       "3            False            False            False            False  \n",
       "4            False            False            False            False  \n",
       "\n",
       "[5 rows x 238 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#for each gene, find the mean of the sum_gene_expr for that gene across all samples\n",
    "\n",
    "# find the means of the sum_gene_expr_normalized\n",
    "gene_sum_normalized_means = combined_data.groupby('gene_name')['sum_gene_expr_normalized'].mean()\n",
    "\n",
    "# find the means of the mean_gene_expr_normalized\n",
    "gene_mean_normalized_means = combined_data.groupby('gene_name')['mean_gene_expr_normalized'].mean()\n",
    "\n",
    "# find the means of the variance_gene_expr_normalized\n",
    "gene_variance_normalized_means = combined_data.groupby('gene_name')['variance_gene_expr_normalized'].mean()\n",
    "\n",
    "# calculate the deviations\n",
    "combined_data['dev_gene_expr_normalized_mean'] = combined_data['gene_name'].map(gene_sum_normalized_means) - combined_data['sum_gene_expr_normalized']\n",
    "combined_data['dev_mean_gene_expr_normalized_mean'] = combined_data['gene_name'].map(gene_mean_normalized_means) - combined_data['mean_gene_expr_normalized']\n",
    "combined_data['dev_variance_gene_expr_normalized_mean'] = combined_data['gene_name'].map(gene_variance_normalized_means) - combined_data['variance_gene_expr_normalized']\n",
    "\n",
    "display(combined_data.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>sum_gene_expr_normalized</th>\n",
       "      <th>mean_gene_expr_normalized</th>\n",
       "      <th>variance_gene_expr_normalized</th>\n",
       "      <th>window_sum_gene_expr_avg</th>\n",
       "      <th>window_mean_gene_expr_avg</th>\n",
       "      <th>window_variance_gene_expr_avg</th>\n",
       "      <th>dev_gene_expr_normalized_mean</th>\n",
       "      <th>dev_mean_gene_expr_normalized_mean</th>\n",
       "      <th>dev_variance_gene_expr_normalized_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>chromosome_chr22</th>\n",
       "      <th>chromosome_chr3</th>\n",
       "      <th>chromosome_chr4</th>\n",
       "      <th>chromosome_chr5</th>\n",
       "      <th>chromosome_chr6</th>\n",
       "      <th>chromosome_chr7</th>\n",
       "      <th>chromosome_chr8</th>\n",
       "      <th>chromosome_chr9</th>\n",
       "      <th>chromosome_chrX</th>\n",
       "      <th>chromosome_chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>2.450584e-09</td>\n",
       "      <td>1.916647e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>4.167977e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>8.853096e-11</td>\n",
       "      <td>0.009572</td>\n",
       "      <td>1.118360e-06</td>\n",
       "      <td>7.938945e-10</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>5.902064e-11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>5.353520e-11</td>\n",
       "      <td>-0.001445</td>\n",
       "      <td>-7.349863e-08</td>\n",
       "      <td>-7.417788e-12</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>4.282816e-11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 234 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      status  sum_gene_expr_normalized  mean_gene_expr_normalized  \\\n",
       "0  amplified                  0.000000               0.000000e+00   \n",
       "1  amplified                  0.006066               4.167977e-07   \n",
       "2  amplified                  0.000000               0.000000e+00   \n",
       "3  amplified                  0.002203               1.510508e-07   \n",
       "4  amplified                  0.000000               0.000000e+00   \n",
       "\n",
       "   variance_gene_expr_normalized  window_sum_gene_expr_avg  \\\n",
       "0                   0.000000e+00                  0.000000   \n",
       "1                   1.770619e-10                  0.003042   \n",
       "2                   0.000000e+00                  0.002028   \n",
       "3                   3.707889e-11                  0.002072   \n",
       "4                   0.000000e+00                  0.001658   \n",
       "\n",
       "   window_mean_gene_expr_avg  window_variance_gene_expr_avg  \\\n",
       "0               0.000000e+00                   0.000000e+00   \n",
       "1               2.083989e-07                   8.853096e-11   \n",
       "2               1.389326e-07                   5.902064e-11   \n",
       "3               1.419621e-07                   5.353520e-11   \n",
       "4               1.135697e-07                   4.282816e-11   \n",
       "\n",
       "   dev_gene_expr_normalized_mean  dev_mean_gene_expr_normalized_mean  \\\n",
       "0                       0.000040                        2.450584e-09   \n",
       "1                       0.009572                        1.118360e-06   \n",
       "2                       0.000000                        0.000000e+00   \n",
       "3                      -0.001445                       -7.349863e-08   \n",
       "4                       0.000000                        0.000000e+00   \n",
       "\n",
       "   dev_variance_gene_expr_normalized_mean  ...  chromosome_chr22  \\\n",
       "0                            1.916647e-12  ...             False   \n",
       "1                            7.938945e-10  ...             False   \n",
       "2                            0.000000e+00  ...             False   \n",
       "3                           -7.417788e-12  ...             False   \n",
       "4                            0.000000e+00  ...             False   \n",
       "\n",
       "   chromosome_chr3  chromosome_chr4  chromosome_chr5  chromosome_chr6  \\\n",
       "0            False            False            False            False   \n",
       "1            False            False            False            False   \n",
       "2            False            False            False            False   \n",
       "3            False            False            False            False   \n",
       "4            False            False            False            False   \n",
       "\n",
       "   chromosome_chr7  chromosome_chr8  chromosome_chr9  chromosome_chrX  \\\n",
       "0            False            False            False            False   \n",
       "1            False            False            False            False   \n",
       "2            False            False            False            False   \n",
       "3            False            False            False            False   \n",
       "4            False            False            False            False   \n",
       "\n",
       "   chromosome_chrY  \n",
       "0            False  \n",
       "1            False  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "\n",
       "[5 rows x 234 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#drop the gene_name column\n",
    "combined_data = combined_data.drop('gene_name', axis=1)\n",
    "\n",
    "#drop the sum_gene_expr, mean_gene_expr, and variance_gene_expr columns, since we are using normalized versions of these columns\n",
    "combined_data = combined_data.drop('sum_gene_expr', axis=1)\n",
    "combined_data = combined_data.drop('mean_gene_expr', axis=1)\n",
    "combined_data = combined_data.drop('variance_gene_expr', axis=1)\n",
    "\n",
    "display(combined_data.head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
