{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Importance Analysis\n",
    "\n",
    "In this notebook, we analyze the feature importances derived from a Random Forest model trained on gene expression data. The dataset contains various features related to gene expression and chromosomal information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>status</th>\n",
       "      <th>sum_gene_expr_normalized</th>\n",
       "      <th>mean_gene_expr_normalized</th>\n",
       "      <th>variance_gene_expr_normalized</th>\n",
       "      <th>window100_sum_gene_expr_avg</th>\n",
       "      <th>window100_mean_gene_expr_avg</th>\n",
       "      <th>window100_variance_gene_expr_avg</th>\n",
       "      <th>window10_sum_gene_expr_avg</th>\n",
       "      <th>window10_mean_gene_expr_avg</th>\n",
       "      <th>window10_variance_gene_expr_avg</th>\n",
       "      <th>...</th>\n",
       "      <th>chromosome_chr22</th>\n",
       "      <th>chromosome_chr3</th>\n",
       "      <th>chromosome_chr4</th>\n",
       "      <th>chromosome_chr5</th>\n",
       "      <th>chromosome_chr6</th>\n",
       "      <th>chromosome_chr7</th>\n",
       "      <th>chromosome_chr8</th>\n",
       "      <th>chromosome_chr9</th>\n",
       "      <th>chromosome_chrX</th>\n",
       "      <th>chromosome_chrY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.006066</td>\n",
       "      <td>4.167977e-07</td>\n",
       "      <td>1.770619e-10</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>8.853096e-11</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>2.083989e-07</td>\n",
       "      <td>8.853096e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>5.902064e-11</td>\n",
       "      <td>0.002028</td>\n",
       "      <td>1.389326e-07</td>\n",
       "      <td>5.902064e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.002203</td>\n",
       "      <td>1.510508e-07</td>\n",
       "      <td>3.707889e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>5.353520e-11</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>1.419621e-07</td>\n",
       "      <td>5.353520e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>amplified</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>4.282816e-11</td>\n",
       "      <td>0.001658</td>\n",
       "      <td>1.135697e-07</td>\n",
       "      <td>4.282816e-11</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      status  sum_gene_expr_normalized  mean_gene_expr_normalized  \\\n",
       "0  amplified                  0.000000               0.000000e+00   \n",
       "1  amplified                  0.006066               4.167977e-07   \n",
       "2  amplified                  0.000000               0.000000e+00   \n",
       "3  amplified                  0.002203               1.510508e-07   \n",
       "4  amplified                  0.000000               0.000000e+00   \n",
       "\n",
       "   variance_gene_expr_normalized  window100_sum_gene_expr_avg  \\\n",
       "0                   0.000000e+00                     0.000000   \n",
       "1                   1.770619e-10                     0.003042   \n",
       "2                   0.000000e+00                     0.002028   \n",
       "3                   3.707889e-11                     0.002072   \n",
       "4                   0.000000e+00                     0.001658   \n",
       "\n",
       "   window100_mean_gene_expr_avg  window100_variance_gene_expr_avg  \\\n",
       "0                  0.000000e+00                      0.000000e+00   \n",
       "1                  2.083989e-07                      8.853096e-11   \n",
       "2                  1.389326e-07                      5.902064e-11   \n",
       "3                  1.419621e-07                      5.353520e-11   \n",
       "4                  1.135697e-07                      4.282816e-11   \n",
       "\n",
       "   window10_sum_gene_expr_avg  window10_mean_gene_expr_avg  \\\n",
       "0                    0.000000                 0.000000e+00   \n",
       "1                    0.003042                 2.083989e-07   \n",
       "2                    0.002028                 1.389326e-07   \n",
       "3                    0.002072                 1.419621e-07   \n",
       "4                    0.001658                 1.135697e-07   \n",
       "\n",
       "   window10_variance_gene_expr_avg  ...  chromosome_chr22  chromosome_chr3  \\\n",
       "0                     0.000000e+00  ...             False            False   \n",
       "1                     8.853096e-11  ...             False            False   \n",
       "2                     5.902064e-11  ...             False            False   \n",
       "3                     5.353520e-11  ...             False            False   \n",
       "4                     4.282816e-11  ...             False            False   \n",
       "\n",
       "   chromosome_chr4  chromosome_chr5  chromosome_chr6  chromosome_chr7  \\\n",
       "0            False            False            False            False   \n",
       "1            False            False            False            False   \n",
       "2            False            False            False            False   \n",
       "3            False            False            False            False   \n",
       "4            False            False            False            False   \n",
       "\n",
       "   chromosome_chr8  chromosome_chr9  chromosome_chrX  chromosome_chrY  \n",
       "0            False            False            False            False  \n",
       "1            False            False            False            False  \n",
       "2            False            False            False            False  \n",
       "3            False            False            False            False  \n",
       "4            False            False            False            False  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#get the combined pckl for the dataframe\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.read_pickle('../combined_data.pkl')\n",
    "\n",
    "display(df.head())\n",
    "\n",
    "#split X and y\n",
    "X = df.drop('status', axis=1)\n",
    "y = df['status']\n",
    "\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "import numpy as np\n",
    "\n",
    "kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "rf = RandomForestClassifier(random_state=42)\n",
    "\n",
    "feature_importances = []\n",
    "\n",
    "for train_index, val_index in kf.split(X):\n",
    "    X_train, X_val = X.iloc[train_index], X.iloc[val_index]\n",
    "    y_train, y_val = y.iloc[train_index], y.iloc[val_index]\n",
    "    \n",
    "    rf.fit(X_train, y_train)\n",
    "    feature_importances.append(rf.feature_importances_)\n",
    "\n",
    "# Calculate mean and standard deviation of feature importances\n",
    "mean_importances = np.mean(feature_importances, axis=0)\n",
    "std_importances = np.std(feature_importances, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>importance</th>\n",
       "      <th>std</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>window100_sum_gene_expr_avg</td>\n",
       "      <td>0.186346</td>\n",
       "      <td>0.000606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>window100_mean_gene_expr_avg</td>\n",
       "      <td>0.177921</td>\n",
       "      <td>0.000594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>window10_sum_gene_expr_avg</td>\n",
       "      <td>0.129314</td>\n",
       "      <td>0.000080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>window10_mean_gene_expr_avg</td>\n",
       "      <td>0.106499</td>\n",
       "      <td>0.000300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dev_gene_expr_normalized_mean</td>\n",
       "      <td>0.102354</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sum_gene_expr_normalized</td>\n",
       "      <td>0.075541</td>\n",
       "      <td>0.000180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>dev_mean_gene_expr_normalized_mean</td>\n",
       "      <td>0.053614</td>\n",
       "      <td>0.000218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mean_gene_expr_normalized</td>\n",
       "      <td>0.048750</td>\n",
       "      <td>0.000188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>chromosome_chr7</td>\n",
       "      <td>0.030213</td>\n",
       "      <td>0.000045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>chromosome_chrX</td>\n",
       "      <td>0.017570</td>\n",
       "      <td>0.000107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>chromosome_chrY</td>\n",
       "      <td>0.012642</td>\n",
       "      <td>0.000118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chromosome_chr10</td>\n",
       "      <td>0.006617</td>\n",
       "      <td>0.000126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chromosome_chr1</td>\n",
       "      <td>0.003749</td>\n",
       "      <td>0.000060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>window100_variance_gene_expr_avg</td>\n",
       "      <td>0.003532</td>\n",
       "      <td>0.000037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chromosome_chr14</td>\n",
       "      <td>0.003257</td>\n",
       "      <td>0.000072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chromosome_chr6</td>\n",
       "      <td>0.002992</td>\n",
       "      <td>0.000077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chromosome_chr5</td>\n",
       "      <td>0.002680</td>\n",
       "      <td>0.000037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chromosome_chr3</td>\n",
       "      <td>0.002568</td>\n",
       "      <td>0.000027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chromosome_chr19</td>\n",
       "      <td>0.002412</td>\n",
       "      <td>0.000067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chromosome_chr12</td>\n",
       "      <td>0.002387</td>\n",
       "      <td>0.000036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>chromosome_chr8</td>\n",
       "      <td>0.002267</td>\n",
       "      <td>0.000037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>chromosome_chr9</td>\n",
       "      <td>0.002211</td>\n",
       "      <td>0.000037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chromosome_chr11</td>\n",
       "      <td>0.002171</td>\n",
       "      <td>0.000019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chromosome_chr2</td>\n",
       "      <td>0.002103</td>\n",
       "      <td>0.000061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>dev_variance_gene_expr_normalized_mean</td>\n",
       "      <td>0.002015</td>\n",
       "      <td>0.000015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chromosome_chr15</td>\n",
       "      <td>0.001946</td>\n",
       "      <td>0.000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chromosome_chr4</td>\n",
       "      <td>0.001915</td>\n",
       "      <td>0.000035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chromosome_chr13</td>\n",
       "      <td>0.001875</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chromosome_chr17</td>\n",
       "      <td>0.001843</td>\n",
       "      <td>0.000063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>window10_variance_gene_expr_avg</td>\n",
       "      <td>0.001818</td>\n",
       "      <td>0.000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chromosome_chr16</td>\n",
       "      <td>0.001804</td>\n",
       "      <td>0.000046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chromosome_chr20</td>\n",
       "      <td>0.001527</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chromosome_chr22</td>\n",
       "      <td>0.001477</td>\n",
       "      <td>0.000018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chromosome_chr18</td>\n",
       "      <td>0.001476</td>\n",
       "      <td>0.000038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chromosome_chr21</td>\n",
       "      <td>0.001325</td>\n",
       "      <td>0.000025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>variance_gene_expr_normalized</td>\n",
       "      <td>0.001266</td>\n",
       "      <td>0.000005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   feature  importance       std\n",
       "3              window100_sum_gene_expr_avg    0.186346  0.000606\n",
       "4             window100_mean_gene_expr_avg    0.177921  0.000594\n",
       "6               window10_sum_gene_expr_avg    0.129314  0.000080\n",
       "7              window10_mean_gene_expr_avg    0.106499  0.000300\n",
       "9            dev_gene_expr_normalized_mean    0.102354  0.000107\n",
       "0                 sum_gene_expr_normalized    0.075541  0.000180\n",
       "10      dev_mean_gene_expr_normalized_mean    0.053614  0.000218\n",
       "1                mean_gene_expr_normalized    0.048750  0.000188\n",
       "31                         chromosome_chr7    0.030213  0.000045\n",
       "34                         chromosome_chrX    0.017570  0.000107\n",
       "35                         chromosome_chrY    0.012642  0.000118\n",
       "13                        chromosome_chr10    0.006617  0.000126\n",
       "12                         chromosome_chr1    0.003749  0.000060\n",
       "5         window100_variance_gene_expr_avg    0.003532  0.000037\n",
       "17                        chromosome_chr14    0.003257  0.000072\n",
       "30                         chromosome_chr6    0.002992  0.000077\n",
       "29                         chromosome_chr5    0.002680  0.000037\n",
       "27                         chromosome_chr3    0.002568  0.000027\n",
       "22                        chromosome_chr19    0.002412  0.000067\n",
       "15                        chromosome_chr12    0.002387  0.000036\n",
       "32                         chromosome_chr8    0.002267  0.000037\n",
       "33                         chromosome_chr9    0.002211  0.000037\n",
       "14                        chromosome_chr11    0.002171  0.000019\n",
       "23                         chromosome_chr2    0.002103  0.000061\n",
       "11  dev_variance_gene_expr_normalized_mean    0.002015  0.000015\n",
       "18                        chromosome_chr15    0.001946  0.000020\n",
       "28                         chromosome_chr4    0.001915  0.000035\n",
       "16                        chromosome_chr13    0.001875  0.000025\n",
       "20                        chromosome_chr17    0.001843  0.000063\n",
       "8          window10_variance_gene_expr_avg    0.001818  0.000010\n",
       "19                        chromosome_chr16    0.001804  0.000046\n",
       "24                        chromosome_chr20    0.001527  0.000016\n",
       "26                        chromosome_chr22    0.001477  0.000018\n",
       "21                        chromosome_chr18    0.001476  0.000038\n",
       "25                        chromosome_chr21    0.001325  0.000025\n",
       "2            variance_gene_expr_normalized    0.001266  0.000005"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#show the importance of each feature\n",
    "importances_df = pd.DataFrame({'feature': X.columns, 'importance': mean_importances, 'std': std_importances})\n",
    "importances_df = importances_df.sort_values('importance', ascending=False)\n",
    "display(importances_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of Feature Importance in Random Forest\n",
    "\n",
    "This table shows the feature importances derived from a Random Forest model. Below is an interpretation of the key columns:\n",
    "\n",
    "## Feature Importance Scores\n",
    "- The `importance` column quantifies the contribution of each feature to the model's predictions. \n",
    "  - **Higher values** indicate more significant features.\n",
    "  - The feature `window100_sum_gene_expr_avg` has the highest importance (0.186346), meaning it plays the most critical role in the model's decision-making.\n",
    "\n",
    "## Standard Deviation (std)\n",
    "- The `std` column represents the variability of the importance scores across all trees in the Random Forest.\n",
    "  - **Lower `std` values** indicate that the importance is consistent across trees.\n",
    "\n",
    "## Key Insights\n",
    "1. **Top Features**:\n",
    "   - Features related to **gene expression averages** (e.g., `window100_sum_gene_expr_avg`, `window100_mean_gene_expr_avg`) are highly influential.\n",
    "   \n",
    "2. **Chromosomal Features**:\n",
    "   - Chromosomal features (e.g., `chromosome_chr7`, `chromosome_chrX`) have relatively lower importance but still contribute slightly to the model.\n",
    "\n",
    "3. **Low-Importance Features**:\n",
    "   - Features like `chromosome_chr8` and `chromosome_chr15` have very low importance values, suggesting they contribute minimally to the predictions.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
